Update PHP and JavaScript dependencies

rhendrickson-sc opened the pull request:
#### Issue

The project had quite a few Dependabot warnings about `gitbook-cli`. We also wanted to update the site overall.

#### Solution

Updated and removed the packages in `composer.json`. Most of the dependencies were updated, however I removed some that weren't being used and/or were causing conflicts/weren't necessary anymore after updating another dependency.

Replaced `gitbook-cli` entirely with [VitePress](https://vitepress.dev). Note: this does have the "negative" that the documentation site will not longer have the "Api References" section because the libraries I could find that would do something like that were either incompatible with VitePress, or were last updated five+ years ago.

#### Impact

This shouldn't have any negative impact. It updates some dependencies and integrates [Rector](https://getrector.com) into the project, to help with this update + future updates.

Note: Rector did add some PHP 8.1 specific changes, but I didn't think this would be a problem considering PHP 8.1 is [end of life in December 2025](https://www.php.net/supported-versions.php).

#### Usage Changes

Only thing I can think of is that these updates do make Conifer a bit more strict on the typing, but I was conservative with it, so we didn't jump from 0% typing to 100% typing.

#### Considerations

At some point we might want to beef up the typing in this project, but that would probably require a new major release + work on any site using Conifer.

#### Testing

All unit tests are passing. No new tests were added.

rbhensley committed:
ran composer update for packages

sdunham committed:
Update composer to prefer stable versions over branches, and update groot dev dependency to install using the master branch to work around wonky release numbers

rhendrickson-sc committed:
deps: remove duplicate json key

rhendrickson-sc committed:
git: update .gitignore

rhendrickson-sc committed:
deps: replace paultehwalton/acf-stubs

Package is no longer maintained, switched to php-stubs/acf-pro-stubs.

rhendrickson-sc committed:
deps: update phpunit/phpunit to v9

rhendrickson-sc committed:
deps: remove acobster/wp-cli-yaml-fixtures

This package is horrendously out of date since Coby left SiteCrafting,
and it isn't being maintained anymore.

Felt like removing the package was the best path forward instead of
forking it and having to update it.

It has a conflict with behat/behat, as they both require `symfony/yaml`,
 but this acobster/wp-cli-yaml-fixtures wants `symfony/yaml:3.4` and the
  newer versions of `behat/behat` want `symfony/yaml:7.x`.

rhendrickson-sc committed:
deps: update `behat/behat`

Fix an "unmaintained packaged" warning when running `lando composer
audit`.

rhendrickson-sc committed:
fix: errors from `lando sniff`

rhendrickson-sc committed:
deps: update szepeviktor/phpstan-wordpress to v2

rhendrickson-sc committed:
fix: PHPStan errors

rhendrickson-sc committed:
deps: set composer.json and .lando.yml to PHP 8.1

rhendrickson-sc committed:
deps: update johnpbloch/wordpress-core

Set it to be ^6.5.5 so it can pull the latest minor version on composer
updates.

rhendrickson-sc committed:
deps: update wp-coding-standards/wpcs to v3

Remove squizlabs/php_codesniffer because wp-coding-standards/wpcs
requires it, so it makes little sense for us to also require it
when we're not using it directly.

rhendrickson-sc committed:
deps: update johnpbloch/wordpress-core-installer

It looks like johnpbloch/wordpress-core-installer isn't really being
maintained anymore, but it was safe to move off the `dev-master` branch
and onto the `2.0.0` branch instead.

The only reason a v2 was released was because the minimum PHP version
got bumped up to PHP 5.6, which doesn't affect us.

rhendrickson-sc committed:
deps: remove gitbook-cli

This package is no longer maintained and no longer works on Node
versions 10 through 22.

After talking to Scott about it, we decided it was better to just strip
it out for now, as it's probably seldom being used.

rhendrickson-sc committed:
deps: remove victorjonsson/markdowndocs

We were only using this to dynamically generate the API Reference
section of the documentation website.
However, this package, and the others like it, are seldom updated, so it
 was more hassle than it was worth to try and keep it.

rhendrickson-sc committed:
deps: replace gitbook-cli with VitePress

Now the documentation site is blazingly fast 🦀

rhendrickson-sc committed:
docs: minor changes

rhendrickson-sc committed:
feat: add Rector

rhendrickson-sc committed:
misc: ran Rector

rhendrickson-sc commented on lib/Conifer/Post/Page.php:
```diff
@@ -29,8 +31,7 @@ class Page extends Post {
   public function get_title_from_nav_or_post( Menu $menu ) : string {
-    return $menu->get_current_top_level_item( $this )->title
-      ?? $this->title();
+    return $menu->get_current_top_level_item()->title ?? $this->title();
```
This is a non-issue. `get_current_top_level_item()` shouldn't have anything passed to it, and it accepts zero arguments.

rhendrickson-sc committed:
fix: CoPilot feedback where it made sense.

sdunham commented on .lando.yml:
```diff
@@ -2,22 +2,25 @@ name: conifer
       - apt-get install subversion -y
 
-    run:
+    build:
```
Can I get some clarification from you on why this was changed from `run` to `build`? I'm running into issues setting up conifer locally, and I'm pretty sure this change is the reason. According to the [Lando docs on build steps](https://docs.lando.dev/services/lando-3.html#build-steps), the `run` step runs after the service has booted up while the `build` step runs before the service has booted up. I think we need to wait to set up WP until after the service has booted up. Is this working on your end?

rhendrickson-sc replied:
Ahh, apologies. I thought I reverted that back after testing some stuff. I'll get a fix pushed in a second for that.

rhendrickson-sc committed:
fix: revert build -> run in .lando.yml

sdunham commented on docs/.vitepress/config.mts:
```diff
@@ -0,0 +1,81 @@
+            {
+                icon: 'x',
+                link: 'https://x.com/sitecrafting',
+            }
```
Tiny thing, but might as well remove this. We don't use twitter anymore.

sdunham commented on docs/index.md:
```diff
@@ -0,0 +1,22 @@
+  - title: Feature B
+    details: Lorem ipsum dolor sit amet, consectetur adipiscing elit
+  - title: Feature C
+    details: Lorem ipsum dolor sit amet, consectetur adipiscing elit
```
Maybe remove or comment this out since we're not using it currently?

rhendrickson-sc committed:
fix: flesh out "features" on the home page

rhendrickson-sc committed:
fix: sniff fixes

There are some lingering sniff "issues" that I don't consider issues.
But the bulk of the problems have been solved.

rhendrickson-sc commented:
@sdunham I made the Sniff updates/fixes. There ended up being two "fixes" I didn't implement.

The first is an issue of `Processing form data without nonce verification`. While it would make sense to fix this, it felt like a breaking change + we'd have to update the documentation. Which felt slightly out of scope for this, but I'm willing to do it if you think it should be done.

The second is an issue of `All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks)`. But I tried fixing those errors with `sanitize_text_field()` and `htmlspecialchars()`, but the issue still got flagged. I'm not sure if it's expecting a WordPress specific function? Also seemed like it might be safe to ignore since the rest of the problems were addressed.

sdunham commented:
@rhendrickson-sc I think you're good to skip those sniff fixes.

I agree that the first one is out of scope for this update. I also agree that implementing it is a good idea, but there will be plenty of edge-cases to account for (especially when you factor in caching, which can cause all kinds of issues when using nonces). I'd say adding a backlog task would be a good idea so it doesn't get lost in the shuffle, but we're definitely going to need to think through the implementation so we don't run into major issues once something like Cloudflare is in the mix.

As for the second one, it's odd that you're still getting a sniff result after adding `sanitize_text_field()` and `htmlspecialchars()`. Feels a bit like a false positive, which is why I think skipping it like you did is fine.
